rfd/0078-role-variables.md

```diff
@@ -0,0 +1,439 @@
+        - out: ['{{regexp.replace(external.username, "-", "_")}}']
+
+        # Values can also take an input, which will usually be a trait.
+        - input: "{{external.email}}"
+          # match is a regular expression which must match the input in order
+          # for this value to be included.
+          match: "^nic@goteleport.com$"
+          out: [root]
+
+        - input: "{{external.email}}"
+          # match may contain contain capture groups which can be expanded in
+          # the output.
+          #
+          # Here "alice@goteleport.com" maps to "alice", but "foo@example.com"
+          # will not be included because it does not match the regex.
+          match: "^(.*)@goteleport.com$"
+          out: ["$1"]
+
+    - name: "allow-env"
+      values:
+        # An input will expand to a list of N inputs if the trait is a list of size N.
+        # Any input value which matches the `match` regex will contribute one
+        # output value, which may contain regex captures from that specific input.
+        #
+        # Here [devs, env-staging, env-prod] maps to [staging, prod].
+        - input: "{{external.groups}}"
+          match: '^env-(\w+)$'
+          out: ['$1']
+
+        # If any of the N input values match the `not_match` regex, none of the
+        # values will be used.
+        #
+        # Here [devs] maps to [dev], but [devs, contractors] maps to []
+        - input: "{{external.groups}}"
+          match: '^devs$'
+          not_match: 'contractors'
+          out: ["dev"]
+  allow:
+    logins:
+      # Variables can be referenced as `vars.<variable_name>` or
+      # `vars["<variable_name"]`.
+      - "{{vars.logins}}"
+    node_labels:
+      env: ["{{vars["allow-env"]}}"]
+```
+
+### How to reference variables
+
+Variables can be referenced similar to traits, by enclosing their name in
+`{{}}`.
+
+Variable names are prefixed by that `vars` namespace to avoid name collisions
+with external traits.
+
+You can use the selector syntax `{{vars.variable_name}}` or the index syntax
+`{{vars["variable-name"]}}`.
+
+It is necessary to use the index syntax and enclose the variable name in quotes
+if it is not a valid
+[Go identifier](https://go.dev/ref/spec#Identifiers),
+or else the parser will not be able to handle it.
+The same is true for traits today.
+
+Variables can only be referenced in the role which defines them.
+The idea of allowing variables to be defined and referenced accross different
+roles was explored, but it has some downsides:
+
+1. Scope of the variable is not clear.
+2. Variable dependencies could create cycles.
+3. Today each role is self-sufficient and can be linted and checked on its own,
+   depending on variables in other roles would break that and make things much
+   more complicated.
+
+### Where variables can be used
+
+Variables can be used anywhere traits can be used in a role spec, including
+allow or deny rules such as `logins` or `node_labels`, impersonation conditions,
+etc.
```

Comment:
What about in label keys?

Reply:
It looks like traits can be used in label keys today, but if the trait expands to a list, only the first value of the list is used (which seems very error-prone). But the var could be used the same way a trait can here.
https://github.com/gravitational/teleport/blob/fc05eaf305f32b943a0e18b3a84c9f61a544b1ae/lib/services/role.go#L446